src/render.c

```diff
@@ -288,57 +351,105 @@ void epd_renderer_deinit() {
+ */
+bool _epd_interlace_line(
+    const uint8_t* to,
+    const uint8_t* from,
+    uint8_t* interlaced,
+    uint8_t* col_dirtyness,
+    int fb_width
+) {
+    uint32_t dirty = 0;
+    int unaligned_start = 0;
+    int unaligned_len = fb_width;
+
+// use Vector Extensions with the ESP32-S3
+#ifdef RENDER_METHOD_LCD
+    // since display horizontal resolutions should be divisible by 16,
+    // only one end should be unaligned by 8 bytes (16 pixels)
+    uint32_t to_addr = (uint32_t)to;
+    int alignment_offset = (to_addr % 16);
+    if (alignment_offset > 0) {
+        unaligned_start = 0;
+        unaligned_len = (16 - alignment_offset) * 2;
+    } else {
+        unaligned_start = fb_width & (~0x1F);
+        unaligned_len = fb_width & 0x1F;
+    }
+
+    dirty = epd_interlace_4bpp_line_VE(
+        to + alignment_offset, from + alignment_offset, interlaced + alignment_offset * 2,
+        col_dirtyness + alignment_offset, fb_width
+    );
+#endif
+    assert((unaligned_start + unaligned_len <= fb_width));
+    for (int x = unaligned_start; x < unaligned_start + unaligned_len; x++) {
+        uint8_t t = *(to + x / 2);
+        uint8_t f = *(from + x / 2);
+        t = (x % 2) ? (t >> 4) : (t & 0x0f);
+        f = (x % 2) ? (f >> 4) : (f & 0x0f);
+        col_dirtyness[x / 2] |= (t ^ f) << (4 * (x % 2));
+        dirty |= (t ^ f);
+        interlaced[x] = (t << 4) | f;
+    }
+
+    return dirty > 0;
+}
 
-    uint8_t* dirty_cols = calloc(epd_width(), 1);
-    assert (dirty_cols != NULL);
+EpdRect epd_difference_image_base(
+    const uint8_t* to,
+    const uint8_t* from,
+    EpdRect crop_to,
+    int fb_width,
+    int fb_height,
+    uint8_t* interlaced,
+    bool* dirty_lines,
+    uint8_t* col_dirtyness
+) {
+    assert(fb_width % 16 == 0);
+    assert(col_dirtyness != NULL);
+
+    memset(col_dirtyness, 0, fb_width / 2);
+    memset(dirty_lines, 0, sizeof(bool) * fb_height);
 
     int x_end = min(fb_width, crop_to.x + crop_to.width);
     int y_end = min(fb_height, crop_to.y + crop_to.height);
 
     for (int y = crop_to.y; y < y_end; y++) {
-        uint8_t dirty = 0;
-        for (int x = crop_to.x; x < x_end; x++) {
-            uint8_t t = *(to + y * fb_width / 2 + x / 2);
-            t = (x % 2) ? (t >> 4) : (t & 0x0f);
-            uint8_t f = *(from + y * fb_width / 2 + x / 2);
-            f = (x % 2) ? (f >> 4) : (f & 0x0f);
-            *from_or |= f;
-            *from_and &= f;
-            dirty |= (t ^ f);
-            dirty_cols[x] |= (t ^ f);
-            interlaced[y * fb_width + x] = (t << 4) | f;
-        }
-        dirty_lines[y] = dirty > 0;
+        uint32_t offset = y * fb_width / 2;
+        int dirty = _epd_interlace_line(
+            to + offset, from + offset, interlaced + offset * 2, col_dirtyness, fb_width
+        );
+        dirty_lines[y] = dirty;
     }
+
     int min_x, min_y, max_x, max_y;
     for (min_x = crop_to.x; min_x < x_end; min_x++) {
-        if (dirty_cols[min_x] != 0)
```

Comment:
Here an advice from my friend Larry Bank who is great optimizing code. Could not this be made faster going 8 pixels at a time? Still have to try this myself but here the idea:

```C
for (; x < x_end; x+=8) { // work 8 pixels (32-bits) at a time
            t = *pTo++; f = *pFrom++;
            dirty |= (t ^ f); // mark this line as dirty of any pixels don't match
            pDirty[x/8] |= (t ^ f); // mark this column as dirty
            // Rest of stuff
            }
```

Reply:
The diff itself in `_epd_interlace_line` should be optimized for speed and there is definitely potential there for the non-vectorized implementation. This particular part is not time-critic though, because it's only done once per update. So I'd rather optimize for legibility here.